### ISSUE TITLE
[Question]: Split long sentences into individually processed chunks to avoid OOM on GPUs

### ISSUE THREAD

**Issue by Guust-Franssens** [flairNLP/flair#3089]
### Problem statement

Whilst fine-tuning [bert-base-multilingual-cased](https://huggingface.co/bert-base-multilingual-cased) I ran into a memory issue on the compute resource. I found this strange since I was on a A100 with 40GB VRAM which should be more than enough to fine-tune this model when using the following settings: 

```python
from flair.embeddings import TransformerWordEmbeddings
from flair.models import SequenceTagger
from flair.trainer import ModelTrainer

embeddings = TransformerWordEmbeddings(
    model="path/to/model",
    layers="-1",
    subtoken_pooling="first",
    fine_tune=True,
    use_context=True,
    allow_long_sentences=True,
    model_max_length=512,
)

tagger = SequenceTagger(
    hidden_size=256, 
    embeddings=embeddings, 
    tag_dictionary=label_dict, 
    tag_type='ner', 
    use_crf=False, 
    use_rnn=False, 
    reproject_embeddings=False, 
) 
  
 # 6. initialize trainer 
 trainer = ModelTrainer(tagger, corpus) 
  
 # 7. run fine-tuning 
 trainer.fine_tune(
    'path/to/save/model', 
    learning_rate=5.0e-6, 
    mini_batch_size=4, 
    mini_batch_chunk_size=1,  
 )
```

I have some very long sentences within my corpus (basically training Flair on documents rather than sentences).

```python
temp = []
for dataset in [corpus.train, corpus.dev, corpus.test]:
   for sentence in dataset:
        temp.append(len(sentence))
print(max(temp))
>>> 14918
```
The longest sentence in my dataset contains 14918 tokens, which I know is a lot...

Filtering out long sentences does not result in OOM which further convinces me of the problem diagnosis. Sentence are split into multiple chunks, but then these chunks are fed to model at the same time instead of split into a chunk for each (which is what you would want if you set a mini_batch_chunk_size=1). Obviously this is faster for the model (larger batch size) but can result in OOM if let's say the sentence is split into a lot of chunks.
```python
14918//512
>>> 29
```
In my case the larger sentence is split into at least 29 chunks (but will be more due to striding).

When I filter out these long sentences I do not run into OOM issues.
```python
corpus.filter_long_sentences(20000)
```

### Solution

Allowing the overflowing chunks to be fed individually to the model instead of all at once, potentially with adding another parameter. e.g.: process_chunk_individually=True

```python
embeddings = TransformerWordEmbeddings(
    model="path/to/model",
    layers="-1",
    subtoken_pooling="first",
    fine_tune=True,
    use_context=True,
    allow_long_sentences=True,
    process_chunk_individually=True, # this slows down the model, but can help prevent OOM
    model_max_length=512,
)
```

### Additional Context

What I describe here can be implemented with Huggingface Transformers like this:

```python
class TokenClassificationChunkPipeline(TokenClassificationPipeline):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def preprocess(self, sentence, offset_mapping=None):
        model_inputs = self.tokenizer(
            sentence,
            return_tensors="pt",
            truncation=True,
            return_special_tokens_mask=True,
            return_offsets_mapping=True,
            return_overflowing_tokens=True,  # Return multiple chunks
            max_length=self.tokenizer.model_max_length,
            padding=True
        )
        if offset_mapping:
            model_inputs["offset_mapping"] = offset_mapping

        model_inputs["sentence"] = sentence

        return model_inputs

    def _forward(self, model_inputs):
        special_tokens_mask = model_inputs.pop("special_tokens_mask")
        offset_mapping = model_inputs.pop("offset_mapping", None)
        sentence = model_inputs.pop("sentence")
        overflow_to_sample_mapping = model_inputs.pop("overflow_to_sample_mapping")

        all_logits = torch.Tensor()
        num_chunks = len(model_inputs["input_ids"])

        # Pass one chunk at a time to the model and concatenate the results
        for i in range(num_chunks):
            model_input = {k: torch.unsqueeze(v[i], dim=0) for k, v in model_inputs.items()}
            logits = model(**model_input)[0]
            all_logits = torch.cat((all_logits, logits), dim=1)

        model_outputs = {
            "logits": all_logits,
            "special_tokens_mask": special_tokens_mask,
            "offset_mapping": offset_mapping,
            "sentence": sentence,
            "overflow_to_sample_mapping": overflow_to_sample_mapping,
            **model_inputs,
        }

        # We reshape outputs to fit with the postprocess inputs
        model_outputs["input_ids"] = torch.reshape(model_outputs["input_ids"], (1, -1))
        model_outputs["token_type_ids"] = torch.reshape(model_outputs["token_type_ids"], (1, -1))
        model_outputs["attention_mask"] = torch.reshape(model_outputs["attention_mask"], (1, -1))
        model_outputs["special_tokens_mask"] = torch.reshape(model_outputs["special_tokens_mask"], (1, -1))
        model_outputs["offset_mapping"] = torch.reshape(model_outputs["offset_mapping"], (1, -1, 2))

        return model_outputs
```

**Comment by helpmefindaname** [comment]
Hi @Guust-Franssens
I have been thinking about your suggestion a bit, however I think this is not really implementable.
While training, the memory usage is often due to intermediate values being saved for backprop, so even with the processing being sequential, we still have to create the same intermediate values, yielding not much less memory while training.

We need all word embeddings for a gradient pass, due to the way models work. (E.g. a text classifier that uses a lstm on top of word embeddings, needs all embeddings).
Note, that the `TokenClassificationChunkPipeline` you provided also won't save the problem you are trying to solve, as it is only about inference and not about training. 

In the case of NER, I would suggest you to use a sentence-splitter, e.g. `sentences = SegtokSentenceSplitter().split("text")` And combine their context:
````
for sent_a, sent_b in zip(sentences[:-1], sentences[1:]):
    sent_a._has_context = True
    sent_a._next_sentence = sent_b
    sent_b._has_context = True
    sent_b._previous_sentence = True
````
And set `use_context=True` when creating the transformer embeddings.

**Note:** you will have to ensure that the labels are aligned right with the newly created sentences, the sentencesplitter will return sentences without label.

That way, you'll have shorter training sentences, but still get the context of the sentences before and after for the individual prediction.

**Comment by Guust-Franssens** [comment]
Hi @helpmefindaname

First of all, I apologize for my late response. I have had a very busy last month, and normally I do not ask a question and then disappear for a month... 

Thanks for your input. Is there some more documentation w.r.t. this SegtokSentenceSplitter() and how to create a custom Flair Corpus with these contextualized sentences? I am very much interested in exploring this further.
